libavcodec/qsvenc_av1.c

```diff
@@ -39,6 +40,73 @@ typedef struct QSVAV1EncContext {
     QSVEncContext qsv;
 } QSVAV1EncContext;
 
+static int qsv_av1_set_encode_ctrl(AVCodecContext *avctx,
+                                   const AVFrame *frame, mfxEncodeCtrl *enc_ctrl)
+{
+    QSVAV1EncContext *q = avctx->priv_data;
+    AVFrameSideData *sd;
+
+    if (!frame || !QSV_RUNTIME_VERSION_ATLEAST(q->qsv.ver, 2, 9))
+        return 0;
+
+    sd = av_frame_get_side_data(frame, AV_FRAME_DATA_MASTERING_DISPLAY_METADATA);
+    if (sd) {
+        AVMasteringDisplayMetadata *mdm = (AVMasteringDisplayMetadata *)sd->data;
+
+        if (mdm->has_primaries && mdm->has_luminance) {
+            const int chroma_den   = 1 << 16;
+            const int max_luma_den = 1 << 8;
+            const int min_luma_den = 1 << 14;
+            int i;
+            mfxExtMasteringDisplayColourVolume *mdcv = av_mallocz(sizeof(mfxExtMasteringDisplayColourVolume));
+
+            if (!mdcv)
+                return AVERROR(ENOMEM);
+
+            mdcv->Header.BufferId = MFX_EXTBUFF_MASTERING_DISPLAY_COLOUR_VOLUME;
+            mdcv->Header.BufferSz = sizeof(*mdcv);
+
+            for (i = 0; i < 3; i++) {
+                mdcv->DisplayPrimariesX[i] =
+                    lrint(chroma_den * av_q2d(mdm->display_primaries[i][0]));
```

Comment:
Don't you want to use av_rescale(mdm->display_primaries[i][0].num, chroma_den, mdm->display_primaries[i][0].den); instead of av_q2d?

Reply:
`av_q2d` and `lrint` were used for display_primaries etc in other places.